modules/data-wrangling/index.qmd

```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
+
+```{r}
+#| eval: false
+penguins |>
+    arrange(bill_length_mm)
+
+penguins |>
+    arrange(desc(bill_length_mm))
+```
+
+```{r}
+#| echo: false
+penguins |>
+    arrange(bill_length_mm) |>
+    head()
+
+penguins |>
+    arrange(desc(bill_length_mm)) |>
+    head()
+```
+
+::: {.callout-tip}
+## Did you know
+
+What do you think will happen if you sort the penguin's data by the species where it is a character variable? If you guess that it will sort in **alphabetical order**, you are correct!!!
+:::
+
+
+## mutate()
+
+This function is for creating new variable or modify the existing one.
+
+```{r}
+#| eval: false
+
+data |>
+    mutate(variable_name = value)
+```
+
+The **"variable_name"**, this can be the existing variable you are going to modify or  new variable name you are trying to create. The **"value"** is what will be assigned to the **"variable_name"**.
+
+Here is an example on using mutata function to modify the **existing** variable
+
+```{r}
+#| eval: false
+penguins |>
+    mutate(sex = ifelse(sex == "male", "M", "F"))
+```
+
+```{r}
+#| echo: false
+penguins |>
+    mutate(sex = ifelse(sex == "male", "M", "F")) |>
+    select(species, sex) |>
+    head()
+```
+
+::: {.callout-caution}
+## Your turn!
+Let said that you think that the body mass of the penguins should be expressed as kilogram instead of gram. How could you create a **new** variable? Make sure you name it as `body_mass_kg`.
+
+Hint: The conversion from gram to kilogram can be expressed in a forms of $kg = g/1000$
+:::
+
+```{webr-teachr}
+# library the packages needed
+library(palmerpenguins)
+library(dplyr)
+
+# Select the column of interests
+penguins |>
+    <<mutate(body_mass_kg = body_mass_g/1000)>>
+
+???
+
+c(
+    "Did you use the right function?" != search_ast(.code, .fn = mutate),
+    "Did you provide the right expression?" != search_ast(.code, .expr = body_mass_g/1000)
+)
+```
+
+
+## group_by()
+
+It allows you to take the existing data and group the specific variables together for future operations. For example, with penguins data, if you want to rank penguins by weights within each species, instead of splitting the dataset by species, rank them individually and then rejoin the data. You can use the `group_by` function together with `mutate` to achieve the same operation.
+
+```{r}
+#| eval: false
+penguins |>
+    group_by(species) |>
+    mutate(ranking = rank(body_mass_g, ties.method = "min")) |>
+    ungroup()
+```
+
+```{r}
+penguins |>
+    group_by(species) |>
+    mutate(ranking = rank(body_mass_g, ties.method = "min")) |>
+    ungroup() |>
+    select(species, body_mass_g, ranking) |>
+    head()
+```
+
+:::{.callout-note}
+It is a good practice to use the `ungroup` function after you have done the operations. If you have not **ungrouped** the data, whatever operations you do after will take into account the grouping whether you want it or not.
```

Comment:
```suggestion
It is a good practice to use the `ungroup()` function after you have done the operations. If you have not **ungrouped** the data, whatever operations you do after will take into account the grouping whether you want it or not.
```

While `ungroup` is technically correct, in my experience new learners need stronger separation of objects and functions. Them being the same is a more advanced topic.